status-im/status-react
# Fix PN metadata leak, PN navigation to chat, and unexpected PNs from other accounts

[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

fixes #6772, #3488 and #2984, depends on https://github.com/status-im/status-go/pull/1297

### Summary:

[comment]: # (Summarise the problem and how the pull request solves it)

Pre-PR example payload:

```javascript
{
	"notification": {
		"title": "Status",
		"body": "You have a new message"
	},
	"data": {
		"msg": {
			"from": "0x04325367620ae20dd878dbb39f69f02c567d789dd21af8a88623dc5b529827c2812571c380a2cd8236a2851b8843d6486481166c39debf60a5d30b9099c66213e4",
			"to": "0x04601f0228b9c046ddf524f7c612093c0ba69a8b7ee37e3f9292626fd6bb50578f57e9f5672e79967fcef7803cfe28daf93342cb4cdb6eb9702eb759d3154b192b"
		}
	}
}
```

Same payload post-PR (notice versioning):
```javascript
{
	"data": {
		"msg-v2": {
			"from": "0x2cea3bd5",
			"to": "0xb1f89744",
			"id": "0x872653ad"
		}
	}
}
```
This PR:
- [x] Adds new message id field to the PN in the form of `hash:${msg-v2/id}`, creating a 1:1 link between Whisper messages and their respective PN on the receiver side 
- [x] Moves to using sender/receiver pubkey hashes instead of the pubkeys themselves (first 10 characters/4 bytes of sha3 hash)
  - [x] Key/values sent: `msg-v2/from`, `msg-v2/to`, `msg-v2/id`
- [x] Stores complete PN payload as JSON in db (`:push-notifications/stored`)
- [x] Moves from Notification API (where title/body were being sent with the PN) to Messaging API (Messaging API only sends a hidden message with key/value pairs) (#6772)
- [x] Maintains compatibility with older clients (receiving only, sent PNs will not be displayed on older clients)
- [x] Creates background task to receive FCM message and display appropriate PN
  - [x] ~~Only create PN if message with hash equal to `msg-v2/id` is not marked as seen~~
- [x] Removes unused `config/in-app-notifications-enabled?`/`IN_APP_NOTIFICATIONS_ENABLED`
- [x] ~~Removes dependency on NotifyUsers status-go API by leveraging react-native-firebase API~~ the API does not support some critical properties for Android and iOS, so we must continue with the implementation in status-go for the time being. Created a suggestion at https://boards.invertase.io/react-native-firebase/p/add-missing-properties-to-messagingremotemessage.
- [x] Fixes navigating to chat if the user is signed out and taps on notification (#3488).

### Review notes (optional):


This PR paves the way to fix issues like https://github.com/status-im/status-react/issues/3488, https://github.com/status-im/status-react/issues/3487, https://github.com/status-im/status-react/issues/2984

### Testing notes (optional):

#### Platforms
Check that new apps can receive PNs from both new and old clients. I've tested Desktop and Android, but not iOS.
- Android (as sender/receiver)
- iOS (as sender/receiver)
- Desktop (as sender only)

See here for expected behavior from a technical perspective: https://github.com/invertase/react-native-firebase-docs/blob/master/docs/messaging/introduction.md#data-only-messages


#### Areas that maybe impacted
**Functional**
- 1-1 chats
- group chats


### Steps to test:
1. Open Status on device 1
1. Open Status on device 2
1. Add device 2 as a contact on device 1
1. Send a message from device 1 to device 2
1. Add device 1 as a contact of device 2
1. Put the app on device 2 in the background
1. Send another message from device 1 to device 2
1. Confirm that a Push Notification is displayed on device 2
1. Tap on the PN
1. Confirm that the app navigates to the appropriate chat



### Future steps

With this PR implemented, we can easily implement the following:

1. Show sender/chat room in PN if db is unlocked (https://github.com/status-im/status-react/issues/7043)
1. Group PNs by sender/chat (https://github.com/status-im/status-react/issues/7046)
1. Navigate to message matching partial IDs in mobile push notifications when PN is tapped #7044

status: ready
